add posibility to output only main dependencies

lewiuberg opened the pull request:
peodd has a much cleaner output than poetry

### Description
add posibility to output only main dependencies
 
### Issues Resolved
n/a
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

vchrombie commented on peodd/peodd.py:
```diff
@@ -28,20 +28,24 @@
 
 @click.command()
 @click.option('-o', '--output',
-              default="requirements-dev.txt",
```
I would remove the `default` argument as a whole but keep it as a required argument, so that the users need to provide the file name.

lewiuberg replied:
```python
@click.command()
@click.option('-o', '--output',
              show_default=True,
              help="Output filename for the dependencies")
```

vchrombie replied:
Since we are not having any default option, I think we can remove the `show_default=True,` too.
WDYT?

lewiuberg replied:
Of course! awkward, haha!

vchrombie commented on peodd/peodd.py:
```diff
@@ -28,20 +28,24 @@
               show_default=True,
               help="Output file for the dependencies")
-def main(output):
+@click.option('-m', '--main',
```
Instead of `main` flag, how about `--not-dev`?

lewiuberg replied:
```python
@click.option('-n', '--not-dev',
              default=False,
              is_flag=True,
              show_default=True,
              help="Output not-dev dependencies")
```

vchrombie replied:
I see there are a lot of non-dev references in the readme and tests, `non-dev` also sounds better than `not-dev`. Do you mind changing it?

Also, I think `-n` is not needed. I would keep it simple
```
@click.option('--non-dev',
```

vchrombie commented on peodd/peodd.py:
```diff
@@ -58,7 +62,10 @@ def main(output):
 
-    poetry_dev_dependencies = toml_dict["tool"]["poetry"]["dev-dependencies"]
+    if main:
+        poetry_dev_dependencies = toml_dict["tool"]["poetry"]["dependencies"]
```
I would rename this variable `poetry_dev_dependencies` since it is not dev_dependencies anymore.
```
-         poetry_dev_dependencies = toml_dict["tool"]["poetry"]["dependencies"]
+         dependencies = toml_dict["tool"]["poetry"]["dependencies"]
```

lewiuberg replied:
```python
dependencies = toml_dict["tool"]["poetry"]["dependencies"]
```

vchrombie commented on peodd/peodd.py:
```diff
@@ -58,7 +62,10 @@ def main(output):
+    else:
+        poetry_dev_dependencies = toml_dict["tool"]["poetry"]["dev-dependencies"]
 
     click.echo("Collected the dev-dependencies")
```
How about changing these statements?

```
    if not_dev:
        dependencies = toml_dict["tool"]["poetry"]["dependencies"]
        click.echo("Collected dependencies ...", nl=False)
    else:
        dependencies = toml_dict["tool"]["poetry"]["dev-dependencies"]
        click.echo("Collected dev-dependencies ...", nl=False)

    click.echo("done")
```

lewiuberg replied:
```python
    if not_dev:
        dependencies = toml_dict["tool"]["poetry"]["dependencies"]
        click.echo("Collected dependencies ...", nl=False)
    else:
        dependencies = toml_dict["tool"]["poetry"]["dev-dependencies"]
        click.echo("Collected dev-dependencies ...", nl=False)

    click.echo("done")
```

vchrombie commented on releases/unreleased/add-support-for-outputting-main-dependencies.yml:
```diff
@@ -0,0 +1,6 @@
+category: added
+author: Lewi Uberg <lewi@uberg.me>
+issue: null
+notes: add posibility to output only main dependencies
```
Do you mind adding a proper note for the feature?
This would eventually go to the release notes and I feel they need to correct and precise.

vchrombie replied:
Good, but I have formatting comment. Will send the suggestion in the next iteration.
Thanks.

vchrombie reviewed (changes requested):
Thank you so much @lewiuberg for opening the PR.

_peodd_ was created with a niche, exporting only the dev-dependencies (untill poetry has this option). I'm open to expand it and add more features if the contributors would find it useful but again, not at the cost of the niche.

I created this tool as I need this specific usecase to install only dev-dependencies and we have integrated this tool with many our workflows. I would be interested to see new features without disturbing the existing working method. Hence, my reviews will be along this line. Thanks for understanding.

Please let me know if you need any help. I can always add the required changes to this PR as new commits if you don't have the time to work on the comments.

Again, thanks a lot for opening a PR to this small awesome tool. :)

vchrombie commented:
Hi @lewiuberg, please don't mind the CI failures. I don't know what is wrong with the coveralls, but I would merge it eventually.

You can continue adding commits to this PR. :)

lewiuberg commented:
Hi @vchrombie!

I will start work on this right away. This will be my first "code review", so bear with me :)

Thanks for the great feedback!

lewiuberg commented:
> Hi @lewiuberg, please don't mind the CI failures. I don't know what is wrong with the coveralls, but I would merge it eventually.
> 
> You can continue adding commits to this PR. :)

It looks like Coveralls is down for maintenance. I use `Codecov` in my package, simple and easy. You can take a look here: https://github.com/lewiuberg/visual-automata/blob/master/.github/workflows/python-package.yml

vchrombie commented on releases/unreleased/add-support-for-outputting-main-dependencies.yml:
````diff
@@ -0,0 +1,6 @@
+category: added
+author: Lewi Uberg <lewi@uberg.me>
+issue: null
+notes: "Add possibility to output only non-dev dependencies to a txt file with the `-n, --non-dev` flag, and remove the default filename, forcing the user to provide one with the `-o, --output` flag. Usage: ``` $ peodd -n -o requirements.txt ```"
````
Looks good but can you add the note in the multiple line format. You can do that by adding a `>` in the starting. Wrap the text to an approximate 72 characters or something.

```
notes: >
    Add possibility to output only non-dev dependencies 
    to a txt file with the `-n, --non-dev` flag, and 
    remove the default filename, forcing the user to 
    provide one with the `-o, --output` flag. 
    Usage: ``` $ peodd -n -o requirements.txt ```
```

vchrombie commented on README.md:
```diff
@@ -71,16 +71,21 @@ Usage: peodd [OPTIONS]
+                     requirements.txt]
 
+  -n, --not-dev      Output not-dev dependencies  [default: False]
   --help             Show this message and exit.
```
Can you update the help manual?

vchrombie commented on tests/test_peodd.py:
```diff
@@ -111,6 +118,31 @@ def test_peodd_script(self, mock_project):
+                text = fd.read()
+
+            self.assertEqual(text, NON_DEV_DEPENDENCIES_CONTENT_REQUIREMENTS_TXT)
+
```
Good work with the tests. :)

vchrombie reviewed (changes requested):
Hi @lewiuberg, I have another round of comments. Please let me know if you need any help.

> I will start work on this right away. This will be my first "code review", so bear with me :)

Sure! :pray: 
Please don't mind, I might be picky sometimes.

> It looks like Coveralls is down for maintenance. I use Codecov in my package, simple and easy. You can take a look here: https://github.com/lewiuberg/visual-automata/blob/master/.github/workflows/python-package.yml

Yeah, I know. Coveralls gets nasty sometimes. I have codecov configured with some other tools but I don't have any solid reason for picking coveralls. XD

lewiuberg commented:
> Hi @lewiuberg, I have another round of comments. Please let me know if you need any help.
> 
> > I will start work on this right away. This will be my first "code review", so bear with me :)
> 
> Sure! 🙏
> Please don't mind, I might be picky sometimes.
> 
> > It looks like Coveralls is down for maintenance. I use Codecov in my package, simple and easy. You can take a look here: https://github.com/lewiuberg/visual-automata/blob/master/.github/workflows/python-package.yml
> 
> Yeah, I know. Coveralls gets nasty sometimes. I have codecov configured with some other tools but I don't have any solid reason for picking coveralls. XD

Please be very picky, I learn from that :)

vchrombie commented on README.md:
```diff
@@ -71,16 +71,21 @@ Usage: peodd [OPTIONS]
-  -o, --output TEXT  Output file for the dependencies  [default: requirements-
-                     dev.txt]
+  -o, --output TEXT  Output file for the dependencies  [default:
+                     requirements.txt]
```
Since we removed the default, I think it should be updated. I have a hack for this.

```
$ poetry install
$ poetry shell
(.venv) $ peodd --help
```

Copy this content and paste it here.

lewiuberg replied:
Usage: peodd [OPTIONS]

  Script to export the pyproject.toml dev-dependencies to a txt file.

Options:
  -o, --output TEXT  Output filename for the dependencies
  --non-dev          Output non-dev dependencies
  --help             Show this message and exit.

vchrombie commented on README.md:
````diff
@@ -71,16 +71,21 @@ Usage: peodd [OPTIONS]
 ```
+Export the non-dev dependencies to `requirements.txt` file
+```
+$ peodd -non-dev -o requirements.txt
````
`-` missing

```
$ peodd --non-dev -o requirements.txt
```

vchrombie commented on peodd/peodd.py:
```diff
@@ -28,20 +28,21 @@
+              help="Output filename for the dependencies")
+@click.option('--non-dev',
+              default=False,
+              is_flag=True,
```
I think you deleted this line in the last commit. I would suggest to remove it for `output` :ballot_box_with_check: but keep it for `non-dev`
```
              show_default=True,
```

lewiuberg replied:
from my last commit:

```python
@click.command()
@click.option('-o', '--output',
              help="Output filename for the dependencies")
@click.option('--non-dev',
              default=False,
              is_flag=True,
              help="Output non-dev dependencies")
```

Shouldn't it be `False`since we want to specifically use `--non-dev` when outputting the non-dev dependencies?

vchrombie replied:
Yes, you are right. :100: 
Sorry for the wrong tip.

vchrombie replied:
Yes, `default=False` and `show_default=True`. It basically shows that default is set to False in the help manual.

vchrombie commented on releases/unreleased/add-support-for-outputting-main-dependencies.yml:
```diff
@@ -0,0 +1,10 @@
+issue: null
+notes: >
+  Add possibility to output only non-dev dependencies to a txt file with the
+  `-n, --non-dev` flag, and remove the default filename, forcing the user to
```
We removed the `-n`. :)

lewiuberg replied:
oh! sorry! it did not show in vscode when I searched the whole project for `-n`.

vchrombie commented on releases/unreleased/add-support-for-outputting-main-dependencies.yml:
````diff
@@ -0,0 +1,10 @@
+  Add possibility to output only non-dev dependencies to a txt file with the
+  `-n, --non-dev` flag, and remove the default filename, forcing the user to
+  provide one with the `-o, --output` flag.
+  Usage: ``` $ peodd -n -o requirements.txt ```
````
same as above, update the same command used in readme

vchrombie reviewed (changes requested):
Thanks for putting the effort @lewiuberg.
We are close!

This is the last set of comments I have in my mind.

lewiuberg commented:
Sorry for all the work! At least you can know that you have passed on some knowledge today :)

vchrombie commented:
> Sorry for all the work! At least you can know that you have passed on some knowledge today :)

No problem, I think we are done with the work.

Do you mind squashing all the commits into one so that I can proceed for merging the PR.
Please let me know if you need any help.

lewiuberg commented:
> > Sorry for all the work! At least you can know that you have passed on some knowledge today :)
> 
> 
> 
> No problem, I think we are done with the work.
> 
> 
> 
> Do you mind squashing all the commits into one so that I can proceed for merging the PR.
> 
> Please let me know if you need any help.

I am not quite sure how this is done, but I will look it up :)

vchrombie commented:
> I am not quite sure how this is done, but I will look it up :)

I can help with that. This is a quick hack.
```
$ git reset --soft HEAD~n && git commit -s
```
where n means the number of commits, here it is 5 for you.

This would open the commit window where you can write your commit message. Once done, you need to force push the branch.
```
$ git push -f origin master
```

It might get messy, but trust me that's how you learn git. First, you fail and then you learn.

vchrombie commented:
> It might get messy, but trust me that's how you learn git. First, you fail and then you learn.

![](https://miro.medium.com/max/600/0*VcMPr1unIjAIHw2j.jpg)

Credits: https://medium.com/@lulu.ilmaknun.q/kompilasi-meme-git-e2fe49c6e33e

lewiuberg commented:
> > It might get messy, but trust me that's how you learn git. First, you fail and then you learn.
> 
> ![](https://camo.githubusercontent.com/c45bbb7f720f8a58bee7ac2b37f4fadabccdfa24072f43ba9d28c1c6abd20a0a/68747470733a2f2f6d69726f2e6d656469756d2e636f6d2f6d61782f3630302f302a56634d507231756e496a41494877326a2e6a7067)
> 
> Credits: https://medium.com/@lulu.ilmaknun.q/kompilasi-meme-git-e2fe49c6e33e

I wish GitHub had a `100%`emoji, haha!

lewiuberg committed:
[peodd] Add support for exporting non-dev dependencies

This commit adds the support for exporting the non-dev dependencies
with the flag `--non-dev`. This is an additional feature, implemented
without tampering the existing working of the tool.

The documentation is updated. Tests are added accordingly.

Signed-off-by: Lewi Uberg <lewiuberg@icloud.com>

lewiuberg commented:
Fingers crossed that I did this right!

vchrombie commented:
> Fingers crossed that I did this right!

Good job @lewiuberg. But, the commit message should something meanigful and precise.

Here is an article you can read when you are free, https://chris.beams.io/posts/git-commit.

You can amend the same commit and edit the commit message. Once you are done, you need to force push the branch again.
```
$ git commit --amend -s
$ git push -f origin master
```

Commit message can be something like this
```
[peodd] Add support for exporting non-dev dependencies

This commit adds the support for exporting the non-dev dependencies
with the flag `--non-dev`. This is an additional feature, implemented
without tampering the existing working of the tool.

The documentation is updated. Tests are added accordingly.

<sign-off>
```

Please let me know if you need any help.

lewiuberg commented:
I read that article a couple of years ago, thanks for finding it for me :)

vchrombie reviewed (approved):
Thanks a lot for working on the changes @lewiuberg. LGTM!

vchrombie commented:
@lewiuberg, thanks for opening the PR and working on the suggestions.

I have released a new version which includes your work. :tada: 
https://github.com/vchrombie/peodd/releases
https://pypi.org/project/peodd

Thanks for contributing.

lewiuberg commented:
Sweet! Thanks for all your help :)

lewiuberg commented:
Hi again!

Maybe the repo header needs to change? Since it's not just for dev-dependence anymore?

vchrombie commented:
Sorry @lewiuberg, I totally missed your comment.

> Maybe the repo header needs to change? Since it's not just for dev-dependence anymore?

Yeah, it makes sense but the main intention of the tool is for dev-dependencies. So, I like to keep it that way. I hope it is fine for you. :handshake:
